pyscf_ipu/experimental/numerics.py

```diff
@@ -0,0 +1,43 @@
+# Copyright (c) 2023 Graphcore Ltd. All rights reserved.
+from functools import wraps
+
+import jax.numpy as jnp
+import numpy as np
+
+
+def apply_fpcast(v, dtype):
+    if isinstance(v, jnp.ndarray) and v.dtype.kind == "f":
+        return v.astype(dtype)
+
+    return v
+
+
+def fpcast(func, dtype=jnp.float32):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        inputs = [apply_fpcast(v, dtype) for v in args]
+        outputs = func(*inputs, **kwargs)
+        return outputs
+
+    return wrapper
+
+
+def compare_fp32_to_fp64(func):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        outputs_fp32 = fpcast(func, dtype=jnp.float32)(*args, **kwargs)
+        outputs_fp64 = fpcast(func, dtype=jnp.float64)(*args, **kwargs)
+        print_compare(outputs_fp32, outputs_fp64)
```

Comment:
Ideally pass `func.__name__`, and print with that.

Reply:
Good idea, I've updated it so the test case now prints:

```
factorial_fori output 0 has max |fp64 - fp32| = 88064.0
factorial_gamma output 0 has max |fp64 - fp32| = 2009088.0
```